### PR TITLE
Fix a few group related issues

### LINKED ISSUE #2334: "Remove entry selection from this group" removes entries from similarly named groups
JabRef 3.7
windows 10 10.0 amd64 
Java 1.8.0_111

Steps to reproduce:

1. Create a Bibtex file with the groups "mygroup" and "mygroup_alternative". Assign a few entries to both of the groups.
2. Select the entries in "mygroup" and remove them from the group using "Remove entry selection from this group".
3. The entries have been removed from both "mygroup" and "mygroup_alternative". Indeed, if you check the group assignments of the entries they now contain "_alternative" as group assignment, i. e., the first part of the group name has been deleted. As no group "_alternative" exists, the items appear to not be assigned to any group.

Note that this bug is related to a workaround/solution to the issue of group hierarchy in JabRef:
https://github.com/JabRef/jabref/issues/1495

If someone was to use

Asthma
    Asthma treatment
    Asthma diagnosis

as proposed in https://github.com/JabRef/jabref/issues/1495, I suspect that the same problem would arise (?).

### PR DESCRIPTION
**Pull request by tobiasdiez**
This PR fixes #2334, #1873 and #1681.
Since I didn't understood the groups code, I also refactored it. Main changes:
- Split `KeywordGroup` class into `WordKeywordGroup`, which splits field content into words and then matches them, and `RegexKeywordGroup` that uses a regular expression for this job.
- Extract serialization of groups to `GroupSerializer` (previously every group decided about its own serialization via the `toString` method)
- Move a few things from model and logic to gui since they mostly contained ui-related stuff (e.g. GroupDesrciptions)

---

- [x] Change in CHANGELOG.md described
- [x] Tests created for changes
- [ ] Screenshots added (for bigger UI changes)
- [x] Manually tested changed features in running JabRef
- [ ] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)
- [ ] If you changed the localization: Did you run `gradle localizationUpdate`?

### CODE REVIEW

**Comment by lenhard** [comment]
The groups code is a mess :-( Thanks for being willing to put up with this!

Unfortunately, the amount of changes is too large for meaningful review and my knowledge on groups and the related code is too limited. I would trust your ability, and the amount of tests that you added here together with the working build. Cosmetic changes can be made on a later occasion. I am sure this is not the last visit to the groups code.

Since this is on your private branch, we have no build on builds.jabref.org. If we had, we could ask the people that reported the issues to try out this version. Any chance of getting a build for this?

File: src/main/java/net/sf/jabref/Groups.uml [hunk 1]
```diff
@@ -0,0 +1,91 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<Diagram>
+  <ID>JAVA</ID>
+  <OriginalElement>net.sf.jabref.model.groups.AbstractGroup</OriginalElement>
```
Comment (koppor) [inline src/main/java/net/sf/jabref/Groups.uml]: Why not using [PlantUML](https://plugins.jetbrains.com/plugin/7017) for documenting?

Which tool did you use?

Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/Groups.uml]: IntelliJ

**Comment by koppor** [comment]
- [ ] dragging an entry to a group does not refresh the main table. (keyword-based group)
- [ ] context menu shows "Remove group, keep subgroups" even though there are no subgroups
- [ ] deletion of a static group does not remove the assignment of entries to it
- [ ] clicking on a group and pressing <kbd>Del</kbd> triggers the dialog "Do you want to remove the selcted entries". If I press <kbd>Cursor Down</kbd>, the next group is selected. Thus <kbd>Del</kbd> should really apply to the current group.

Nothing important or me. Think, this was also happening using the old groups code.

**Pull request merged**